### DISCUSSION TITLE
How to use KaTeX in next.js

### DISCUSSION THREAD

**Discussion by anonymous-qsh** [vercel/next.js#15479] [Help]
How to use KaTex in next.js.

**Comment by iammarkps** [comment]
If you are using MDX you can do like this.

```js
const withMDX = require('@next/mdx')({
  extension: /\.mdx?$/,
  options: {
    remarkPlugins: [remarkMath],
    rehypePlugins: [rehypeKatex, rehypePrism],
  },
})
```

Or I our case my team use [this](https://github.com/programming-in-th/programming.in.th/blob/dev/src/lib/renderMarkdown.ts) method and add KaTeX css manually.

We use that function in here https://github.com/programming-in-th/programming.in.th/blob/dev/src/pages/tasks/%5B...id%5D.tsx

**Reply by iammarkps** [reply]
I don't know if its the best approach, feel free to tell me if you have a better idea!

**Reply by mysterybear** [reply]
Your link is now like https://github.com/programming-in-th/programming.in.th/blob/main/src/lib/renderMarkdown.ts and I don't see any katex css here?

I've got it working the standard way recommended, something like:

```tsx
const PostPage = ({ source, frontMatter }: Props) => {
  return (
    <Layout>
      <Head>
        <link
          rel="stylesheet"
          href="https://cdn.jsdelivr.net/npm/katex@0.13.11/dist/katex.min.css"
          integrity="sha384-Um5gpz1odJg5Z4HAmzPtgZKdTBHZdw8S29IecapCSB31ligYPhHQZMIlWLYQGVoc"
          crossOrigin="anonymous"
        />
      </Head>
      <header>
        <nav>
          <Link href="/">
            <a>👈 Go back home</a>
          </Link>
        </nav>
      </header>
      <div className="post-header">
        <h1>{frontMatter.title}</h1>
        {frontMatter.description && (
          <p className="description">{frontMatter.description}</p>
        )}
      </div>
      <main>
        <MDXRemote {...source} components={components} />
      </main>

      <style jsx>{`
        .post-header h1 {
          margin-bottom: 0;
        }

        .post-header {
          margin-bottom: 2rem;
        }
        .description {
          opacity: 0.6;
        }
      `}</style>
    </Layout>
  )
}
```

But I get a FOUC on loading, wondering if there's a better way to load the katex css? Thought about just looking at the source and then re-writing it as CSS modules...

**Reply by elieven** [reply]
How exactly are you importing `remarkMath`, `rehypeKatex` and `rehypePrism`? I'm getting errors if I `const rehypeKatex = require('rehype-katex')` or `import rehypeKatex from 'rehype-katex'`.  I'm not too sure what the difference is with `const rehypeKatex = import('rehype-katex')` and while that one does not give me errors it does not actually work / do anything to any pages with math inside (does not work). I've included all page extensions in `pageExtensions` that I use as well as tried every possible surrounding characters that surround math eg. `$$math$$`, `\(math\)`... and none of it works.

**Reply by erickcestari** [reply]
Yes, I know it's been 2 years, but just posting here how it worked for me using the router app, in case anyone needs it.

```
import nextMDX from "@next/mdx";
import rehypeHighlight from "rehype-highlight";
import rehypeKatex from 'rehype-katex'
import remarkMath from 'remark-math'

const nextConfig = {
  pageExtensions: ['js', 'jsx', 'mdx', 'ts', 'tsx'],
  images: {
    remotePatterns: [
      {
        protocol: 'https',
        hostname: 'avatars.githubusercontent.com',
      },
    ],
  },
}

const withMDX = nextMDX({
  options: {
    remarkPlugins: [remarkMath],
    rehypePlugins: [rehypeHighlight, rehypeKatex],
  },
});

export default withMDX(nextConfig);
```
